core/chains/evm/logpoller/orm.go

```diff
@@ -240,11 +255,30 @@ type Exp struct {
 	ShouldDelete bool
 }
 
-func (o *DbORM) DeleteExpiredLogs(qopts ...pg.QOpt) error {
+func (o *DbORM) DeleteExpiredLogs(limit int64, qopts ...pg.QOpt) (int64, error) {
 	qopts = append(qopts, pg.WithLongQueryTimeout())
 	q := o.q.WithOpts(qopts...)
 
-	return q.ExecQ(`WITH r AS
+	if limit > 0 {
+		return q.ExecQWithRowsAffected(`
+		DELETE FROM evm.logs
+		WHERE (block_hash, log_index, evm_chain_id) IN (
```

Comment:
Use primary key for nested query

Reply:
@reductionista Inspired by your DM about primary_key usage here, I did another round of tests and discovered that using another index is faster. It's mostly because primary_key has different fields than the index used for nested query. It's faster because it allows PostgreSQL to quickly locate the rows to be deleted based on the evm_chain_id, block_number, address, and event_sig columns.

Primary key search
```sql
begin;
explain analyze DELETE FROM evm.logs
                WHERE (block_hash, log_index, evm_chain_id) IN (
                        SELECT l.block_hash, l.log_index, l.evm_chain_id
                        FROM evm.logs l
                        INNER JOIN (
                                SELECT address, event, 10 AS retention
                                FROM evm.log_poller_filters
                                WHERE evm_chain_id = 11155111
                        ) r ON l.evm_chain_id = 11155111 AND l.address = r.address AND l.event_sig = r.event
                        AND l.created_at <= STATEMENT_TIMESTAMP() - (r.retention / 10^9 * interval '1 second')
                        LIMIT 10000
                );
rollback;

 Delete on logs  (cost=11491.49..89413.65 rows=1049402 width=77) (actual time=128.404..128.407 rows=0 loops=1)
   ->  Nested Loop  (cost=11491.49..89413.65 rows=1049402 width=77) (actual time=22.917..94.569 rows=10000 loops=1)
         ->  HashAggregate  (cost=11490.93..11590.93 rows=10000 width=118) (actual time=22.891..26.820 rows=10000 loops=1)
               Group Key: "ANY_subquery".block_hash, "ANY_subquery".log_index, "ANY_subquery".evm_chain_id
               Batches: 1  Memory Usage: 3217kB
               ->  Subquery Scan on "ANY_subquery"  (cost=0.56..11415.93 rows=10000 width=118) (actual time=5.080..18.151 rows=10000 loops=1)
                     ->  Limit  (cost=0.56..11315.93 rows=10000 width=47) (actual time=5.075..15.761 rows=10000 loops=1)
                           ->  Nested Loop  (cost=0.56..216604.91 rows=191425 width=47) (actual time=5.074..14.994 rows=10000 loops=1)
                                 ->  Seq Scan on log_poller_filters  (cost=0.00..56.16 rows=295 width=54) (actual time=0.009..0.062 rows=49 loops=1)
                                       Filter: (evm_chain_id = '11155111'::numeric)
                                       Rows Removed by Filter: 102
                                 ->  Index Scan using idx_evm_logs_ordered_by_block_and_created_at on logs l  (cost=0.56..730.05 rows=401 width=101) (actual time=0.101..0.272 rows=204 loops=49)
                                       Index Cond: ((evm_chain_id = '11155111'::numeric) AND (address = log_poller_filters.address) AND (event_sig = log_poller_filters.event) AND (created_at <= (statement_timestamp() - '00:00:00'::interval)))
         ->  Index Scan using logs_pkey on logs  (cost=0.56..8.40 rows=1 width=53) (actual time=0.006..0.006 rows=1 loops=10000)
               Index Cond: ((block_hash = "ANY_subquery".block_hash) AND (log_index = "ANY_subquery".log_index) AND (evm_chain_id = "ANY_subquery".evm_chain_id))
 Planning Time: 0.791 ms
 Execution Time: 128.919 ms
(17 rows)
```

INdex
```sql
begin;
explain analyze DELETE FROM evm.logs
        WHERE  (evm_chain_id, block_number, address, event_sig) IN (
                        SELECT l.evm_chain_id, l.block_number, l.address, l.event_sig
                        FROM evm.logs l
                        INNER JOIN (
                                SELECT address, event, 10 AS retention
                                FROM evm.log_poller_filters
                                WHERE evm_chain_id = 11155111
                        ) r ON l.evm_chain_id = 11155111 AND l.address = r.address AND l.event_sig = r.event
                        AND l.created_at <= STATEMENT_TIMESTAMP() - (r.retention / 10^9 * interval '1 second')
                        LIMIT 10000);
rollback;

 Delete on logs  (cost=797.00..85262.44 rows=1050566 width=98) (actual time=83.777..83.779 rows=0 loops=1)
   ->  Nested Loop  (cost=797.00..85262.44 rows=1050566 width=98) (actual time=23.875..52.695 rows=10001 loops=1)
         ->  HashAggregate  (cost=796.44..896.44 rows=10000 width=160) (actual time=23.851..24.835 rows=2676 loops=1)
               Group Key: "ANY_subquery".evm_chain_id, "ANY_subquery".block_number, "ANY_subquery".address, "ANY_subquery".event_sig
               Batches: 1  Memory Usage: 1169kB
               ->  Subquery Scan on "ANY_subquery"  (cost=0.56..696.44 rows=10000 width=160) (actual time=5.045..18.784 rows=10000 loops=1)
                     ->  Limit  (cost=0.56..596.44 rows=10000 width=68) (actual time=5.040..16.282 rows=10000 loops=1)
                           ->  Nested Loop  (cost=0.56..11407.16 rows=191425 width=68) (actual time=5.039..15.512 rows=10000 loops=1)
                                 ->  Seq Scan on log_poller_filters  (cost=0.00..56.16 rows=295 width=54) (actual time=0.010..0.061 rows=49 loops=1)
                                       Filter: (evm_chain_id = '11155111'::numeric)
                                       Rows Removed by Filter: 102
                                 ->  Index Only Scan using idx_evm_logs_ordered_by_block_and_created_at on logs l  (cost=0.56..34.47 rows=401 width=68) (actual time=0.101..0.293 rows=204 loops=49)
                                       Index Cond: ((evm_chain_id = '11155111'::numeric) AND (address = log_poller_filters.address) AND (event_sig = log_poller_filters.event) AND (created_at <= (statement_timestamp() - '00:00:00'::interval)))
                                       Heap Fetches: 10000
         ->  Index Scan using evm_logs_idx on logs  (cost=0.56..8.43 rows=1 width=74) (actual time=0.008..0.009 rows=4 loops=2676)
               Index Cond: ((evm_chain_id = "ANY_subquery".evm_chain_id) AND (block_number = "ANY_subquery".block_number) AND (address = "ANY_subquery".address) AND (event_sig = "ANY_subquery".event_sig))
 Planning Time: 0.763 ms
 Execution Time: 84.091 ms
(18 rows)
```